lib/logsource.c

```diff
@@ -169,6 +197,153 @@ log_source_flow_control_suspend(LogSource *self)
   window_size_counter_suspend(&self->window_size);
 }
 
+void
+log_source_enable_dynamic_window(LogSource *self, DynamicWindowPool *window_ctr)
+{
+  dynamic_window_set_counter(&self->dynamic_window, dynamic_window_pool_ref(window_ctr));
+}
+
+void
+log_source_dynamic_window_update_statistics(LogSource *self)
+{
+  dynamic_window_stat_update(&self->dynamic_window.stat, window_size_counter_get(&self->window_size, NULL));
+  msg_trace("Updating dynamic window statistic", evt_tag_int("avg window size",
+                                                             dynamic_window_stat_get_avg(&self->dynamic_window.stat)));
+}
+
+static void
+_reclaim_dynamic_window(LogSource *self, gsize window_size)
+{
+  g_assert(self->full_window_size - window_size >= self->options->init_window_size);
+  atomic_gssize_set(&self->window_size_to_be_reclaimed, window_size);
+}
+
+static void
+_release_dynamic_window(LogSource *self)
+{
+  g_assert(self->ack_tracker == NULL);
+
+  gsize dynamic_part = self->full_window_size - self->options->init_window_size;
+  msg_trace("Releasing dynamic part of the window", evt_tag_int("dynamic_window_to_be_released", dynamic_part),
+            log_pipe_location_tag(&self->super));
+  self->full_window_size -= dynamic_part;
+  window_size_counter_sub(&self->window_size, dynamic_part, NULL);
+  dynamic_window_release(&self->dynamic_window, dynamic_part); //TODO: rename release to ...
+
+  dynamic_window_pool_unref(self->dynamic_window.ctr); //TODO: move to dynamic_window_release_counter()
+}
+
+static void
+_inc_balanced(LogSource *self, gsize inc)
+{
+  gsize offered_dynamic = dynamic_window_request(&self->dynamic_window, inc);
+
+  msg_warning("Balance::increase",
+              log_pipe_location_tag(&self->super),
+              evt_tag_printf("connection", "%p", self),
+              evt_tag_int("old_full_window_size", self->full_window_size),
+              evt_tag_int("new_full_window_size", self->full_window_size + offered_dynamic));
+
+  self->full_window_size += offered_dynamic;
+
+  gsize old_window_size = window_size_counter_add(&self->window_size, offered_dynamic, NULL);
+  if (old_window_size == 0 && offered_dynamic != 0)
+    log_source_wakeup(self);
+}
+
+static void
+_dec_balanced(LogSource *self, gsize dec)
+{
+  gsize new_full_window_size = MAX(self->options->init_window_size, dec);
+  gsize subtrahend = self->full_window_size - new_full_window_size;
+
+  gsize empty_window = window_size_counter_get(&self->window_size, NULL);
+  gsize remaining_sub = 0;
+  if (empty_window <= subtrahend)
+    {
+      remaining_sub = subtrahend - empty_window;
+      if (empty_window == 0)
+        {
+          subtrahend = 0;
+        }
+      else
+        {
+          subtrahend = empty_window - 1;
```

Comment:
Where does this `1` come from? Shouldn't this be the `static-window-size`?

Reply:
`- 1` is here to avoid suspending the source on one more code path. The new `window_size` would be 0 without `- 1`).

The suspend-wakeup logic looks simple, but its task ordering and threading implications  are complicated enough, that's why I did not want to add another suspend call.

Reply:
Ok, thanks! :)